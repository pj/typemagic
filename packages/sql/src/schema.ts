


function validateSchema() {

}

export type ComputeSchema<From, Join> =
  {

  }