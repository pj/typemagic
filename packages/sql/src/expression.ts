
export type ValidateStringExpression<Schema, Exp> =
  any