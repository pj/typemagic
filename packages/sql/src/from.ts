// Calculate column names of from/joins necessary for select/where/group/order etc.

type ComputeAllIdentifiers<From, Identifier> =
  Identifier extends infer X ? `${string & From}__${string & X}` : never

type ComputeAllJoins<Schema, Join> =
  Join extends {table: infer Table, as: infer As}
    ? Table extends keyof Schema
      ? unknown extends As
        ? ComputeAllIdentifiers<Table, Schema[Table]>
        : ComputeAllIdentifiers<As, Schema[Table]>
      : never
    : never

export type ComputeFrom<Schema, From> =
    [From] extends [string]
      ? From extends keyof Schema
        ? ComputeAllIdentifiers<From, Schema[From]>
        : never
      : From extends [infer OldName, infer AliasName]
        ? OldName extends string
          ? AliasName extends string
            ? OldName extends keyof Schema
              ? ComputeAllIdentifiers<AliasName, Schema[OldName]>
              : never
            : [keyof Schema, string]
          : [keyof Schema, string]
        : never

export type ComputeFromAndJoins<Schema, From, Joins> =
  [unknown] extends [From]
    ? [unknown] extends [Joins]
      ? ComputeFrom<Schema, From>
      : ComputeFrom<Schema, From> | ComputeAllJoins<Schema, Joins>
    : unknown
