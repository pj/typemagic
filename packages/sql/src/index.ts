
type ValidateSelect<Select, DefaultTable, JoinTables> =
  [Select] extends [Array<'*' | keyof DefaultTable | [DefaultTable, keyof DefaultTable]> | {table: infer TableName, column: infer ColumnName}]
    ?  Select
    : 'Unable to figure out select'

type ValidateTableSchema<TableSchema> =
  [TableSchema] extends [{
    source: infer Source,
    nullable?: infer Nullable 
    type: infer Type
  }]
  ? "asdf"
  : "sadf"


type ExtractDefaultTable<From> =
  [From] extends [{
    table: infer Table,
  }]
    ? Table
    : [From] extends [infer Table]
      ? Table
      : unknown

export enum JoinTypes {
  FullOuterJoin,
  InnerJoin,
  LeftJoin
}


type ExtractJoinTable<Join, DefaultTable> =
  Join extends {
    table: infer Table 
    joinType?: infer JoinType
    on: infer On
  } 
    ? {
        table: Table,
        joinType: JoinType,
        on: undefined
      }
    : Join extends [infer SourceTable, infer X, infer DestTable, infer Y]
      ? [SourceTable, keyof SourceTable, DestTable, keyof DestTable]
      : "asdf"

type ExtractJoinTables<From, DefaultTable> =
  From extends {
    joins?: infer Joins
  }
    ? ExtractJoinTable<Joins, DefaultTable>
    : unknown

type ValidateFrom<From> =
  [From] extends [{
    table: infer Table,
    join?: infer Joins
  }]
    ? "here"
    : 'asdf'


type ValidateQuery<Query> =
  [Query] extends [
    {
      select?: infer Select,
      from?: infer From 
    }
  ] 
    ? ValidateSelect<
        Select, 
        ExtractDefaultTable<From>, 
        ExtractJoinTables<From, ExtractDefaultTable<From>>
      >
      & ValidateFrom<From>
   : "Can't infer query"


export function query<Query extends ValidateQuery<Query>>(query: Query) {

}