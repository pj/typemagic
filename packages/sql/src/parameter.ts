export class Parameter<Type> {
  constructor (public position: number | null ) {}
}

// export const parameter = new Parameter(null);

export function param<Type = unknown>(position: number | null) {
  return new Parameter<Type>(position);
}