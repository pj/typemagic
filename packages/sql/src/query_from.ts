type ValidateSchema<Schema> =
  {
    [Key in keyof Schema]?:
      "Asdf"
  }

// Get column names that are common across the various tables that are being joined
type GetDuplicateColumns<Schemas> = 
  any

// Get Valid Column names for select.
type GetColumnTypes<From, Join, GroupNames> = 
  any

// Get all schemas
type GetSchemas<From, Joins> =
  any


type ValidateSelect<ColumnTypes> = {

}

type ValidateFrom<From> =
  [From] extends [{
    schema: infer Schema,
    join?: infer Joins
  }]
    ? "here"
    : 'asdf'

type ValidateJoin<Join> =
  [Join] extends [{
    schema: infer Schema,
    join?: infer Joins
  }]
    ? "here"
    : 'asdf'

type ValidateWhere<Where> =
  [Where] extends [{
    schema: infer Schema,
    join?: infer Joins
  }]
    ? "here"
    : 'asdf'

type ValidateQuery<Query> =
  [Query] extends [{
    select: infer Select,
    from?: infer From,
    join?: infer Join,
    where?: infer Where,
    orderBy?: infer OrderBy,
    groupBy?: infer GroupBy
  }]
    ? ValidateSelect<GetColumnTypes<From, Join, GroupBy>>
      & ValidateFrom<From>
    : "Unable to validate query"

