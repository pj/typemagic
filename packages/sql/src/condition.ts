import { ValidateExpression } from "./expression";
import { Parameter } from "./parameter";
import {format} from "@scaleleap/pg-format";

type ValidateLogicalOperators<Schema, Identifiers, Condition> =
  Condition extends {not: infer Not}
    ? {not: ValidateBoolean<Schema, Identifiers, Not>}
    : Condition extends {and: infer And}
      ? And extends Readonly<unknown[]>
        ? {and: ValidateBoolean<Schema, Identifiers, And>}
        : {and: "Must be an array"}
      : Condition extends {or: infer Or}
        ? Or extends Readonly<unknown[]>
          ? {or: ValidateBoolean<Schema, Identifiers, Or>}
          : {or: "Must be an array"}
        : never

type ValidateComparators = 
  'eq' | 'ne' | 'equal' | 'not_equal' 
  | 'lt' | 'lte' | 'less_than' | 'less_than_equal' 
  | 'gt' | 'gte' | 'greater_than' | 'greater_than_equal' 
  | 'in' | 'not in';

type ValidateEqualityOperators<Schema, Condition> =
  {
    [Key in keyof Condition]: 
      [Key] extends ValidateComparators
        ? Condition[Key] extends [infer Left, infer Right]
          ? [ValidateExpression<Schema, Left>, ValidateExpression<Schema, Right>]
          : ["Must be two element tuple", "Must be two element tuple"]
        : never
  } extends infer Validated
    ? Validated extends {[Key in keyof Condition]: never}
      ? never
      : Validated
    : never

export type ValidateBoolean<Schema, Identifiers, Condition> =
  // Literals
  Condition extends boolean
    ? boolean
    // Parameters
    : Condition extends Parameter<infer Type>
      ? Type extends boolean
        ? Parameter<Type>
        : "Parameter must be boolean"
      // Identifiers
      : Condition extends Identifiers
        ? Condition
        // Logical Operators
        : ValidateLogicalOperators<Schema, Identifiers, Condition> extends never
          // Comparators
          ? ValidateEqualityOperators<Schema, Condition> extends never
            ? "Unable to validate - must be boolean literal/operator/parameter"
            : ValidateEqualityOperators<Schema, Condition>
          : ValidateLogicalOperators<Schema, Identifiers, Condition>
      
    // TODO: Boolean functions

export function generateCondition(where: any): string {
  if (Array.isArray(where)) {
    return where.map(w => generateCondition(w)).join(" AND ");
  } else if (where["not"]) {
    return ` NOT (${generateCondition(where["not"])})`
  } else if (where === true || where === false) {
    return where ? "TRUE" : "FALSE"
  }

  throw new Error("Unknown condition");
}
