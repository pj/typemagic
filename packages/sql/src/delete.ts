// DELETE FROM [ ONLY ] table_name [ * ] [ [ AS ] alias ]
//     [ USING from_item [, ...] ] -- not standard
//     [ WHERE condition | WHERE CURRENT OF cursor_name ]
//     [ RETURNING * | output_expression [ [ AS ] output_name ] [, ...] ]
import {format} from "@scaleleap/pg-format";
import {format as syntaxFormat} from "sql-formatter";
import { generateCondition, ValidateBoolean } from "./condition";
import { ComputeFrom } from "./from";
import { ValidateSchema, ValidateTableName } from "./schema";
import dedent from 'dedent-js'

export type ValidateWhere<Schema, Identifiers, Where> =
  [unknown] extends [Where]
    ? {}
    : {where: ValidateBoolean<Schema, Identifiers,  Where>}

export type ValidateDelete<Schema, Delete> = 
  [Delete] extends [{
    from: infer TableName,
    only?: infer Only,
    where?: infer Where,
    returning?: infer Returning
  }]
    ?  ValidateTableName<Schema, TableName> 
    //   // & unknown extends Only ? {} : {only: boolean}
      & ValidateWhere<Schema, ComputeFrom<Schema, TableName>, Where>
    : "Unable to validate delete"

export function generateTableName(from: any) {
  if (Array.isArray(from)) {
    let [tableName, aliasName] = from;
    return format("%I AS %I", tableName, aliasName);
  } else {
    return format("%I", from);
  }
}

export function remove<
  Schema extends ValidateSchema<Schema>, 
  Delete extends ValidateDelete<Schema, Delete>
>(schema: Schema, remove: Delete): string {
  const removeAny = remove as any; 
  return syntaxFormat(`
    DELETE FROM ${removeAny.only ? "ONLY" : ""} ${generateTableName(removeAny.from)}
    WHERE ${generateCondition(removeAny.where)}
  `);
}