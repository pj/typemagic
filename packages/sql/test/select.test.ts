import { ArrayToUnion, DuplicatedColumns, GenerateColumnNames, select, ValidateSelect } from '../src';

const schema = {
  'test': [
    'id', 'name', 'description'
  ], 
  'other': [
    'id', 'name', 'thing', 'test_id'
  ],
  'another': [
    'id', 'date', 'test_id'
  ]
} as const;

type X = 
  ValidateSelect<
    typeof schema, 
    ['test', 'renamed'], 
    [
      ['other', 'test.id', 'other.test_id']
    ],
    [
      ['renamed.id', 'renamed_id'], 
      ['other.id', 'other_id'],
      ['other.name', 'name'],
      'thing',
      'description'
    ]
  >

type D = DuplicatedColumns<
  typeof schema, 
  ['test', 'renamed'], 
  [
    ['other', 'other.test_id', 'test.id'],
    ['another', 'another.test_id', 'test.id']
  ]
>

type C = GenerateColumnNames<'test', 'renamed', ArrayToUnion<typeof schema, 'test'>, 'id' | never>

test(
  'select', 
  async () => {
    expect(
      select(
        schema,
        {
          select: [
            ['renamed.id', 'renamed_id'], 
            ['other.id', 'other_id'],
            ['other.name', 'name'],
            'thing',
            'description'
          ], 
          from: ['test', 'renamed'],
          join: [
            ['other', 'test.id', 'other.test_id']
          ]
        } as const
      )
    ).toMatchSnapshot();
  }
);