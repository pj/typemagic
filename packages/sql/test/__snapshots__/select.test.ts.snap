// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select 1`] = `
"SELECT
FROM
  test AS renamed"
`;
